statistics/statistics.go

```diff
@@ -212,94 +249,150 @@ func (c *Column) mergeBuckets(bucketIdx int64) {
-		Values:  make([]types.Datum, len(values)),
-		Repeats: cpb.GetRepeats(),
+	selSQL := fmt.Sprintf("select bucket_id, count, repeats, value from mysql.stats_buckets where table_id = %d and %s = %d", tableID, colName, colID)
+	rows, _, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, selSQL)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
-	for i, val := range values {
-		c.Values[i], err = tablecodec.Unflatten(val, ft, false)
-		if err != nil {
-			return nil, errors.Trace(err)
+	bucketSize := len(rows)
+	if bucketSize == 0 {
+		return nil, nil
+	}
+	colStats := &Column{
+		ID:      colID,
+		NDV:     distinct,
+		Numbers: make([]int64, bucketSize),
+		Repeats: make([]int64, bucketSize),
+		Values:  make([]types.Datum, bucketSize),
+	}
+	for i := 0; i < bucketSize; i++ {
+		bucketID := rows[i].Data[0].GetInt64()
+		count := rows[i].Data[1].GetInt64()
+		repeats := rows[i].Data[2].GetInt64()
+		var value types.Datum
+		if isIndex {
+			value = rows[i].Data[3]
+		} else {
+			value, err = rows[i].Data[3].ConvertTo(ctx.GetSessionVars().StmtCtx, tp)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
 		}
+		colStats.Numbers[bucketID] = count
+		colStats.Repeats[bucketID] = repeats
+		colStats.Values[bucketID] = value
+	}
+	for i := 1; i < bucketSize; i++ {
+		colStats.Numbers[i] += colStats.Numbers[i-1]
 	}
-	return c, nil
+	return colStats, nil
 }
 
-// Table represents statistics for a table.
-type Table struct {
-	Info    *model.TableInfo
-	TS      int64 // build timestamp.
-	Columns []*Column
-	Indices []*Column
-	Count   int64 // Total row count in a table.
-	Pseudo  bool
+// TableStatsFromStorage load table stats info from storage.
+func TableStatsFromStorage(ctx context.Context, info *model.TableInfo, count int64) (*Table, error) {
+	table := &Table{
+		Info:  info,
+		Count: count,
+	}
+	selSQL := fmt.Sprintf("select * from mysql.stats_columns where table_id = %d", info.ID)
+	rows, _, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, selSQL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, row := range rows {
+		distinct := row.Data[3].GetInt64()
+		if row.Data[1].IsNull() {
+			// process index
+			colID := row.Data[2].GetInt64()
+			col, err := colStatsFromStorage(ctx, info.ID, colID, nil, distinct, true)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			table.Indices = append(table.Indices, col)
+		} else {
+			// process column
+			colID := row.Data[1].GetInt64()
+			var col *Column
+			for _, colInfo := range info.Columns {
+				if colID == colInfo.ID {
+					col, err = colStatsFromStorage(ctx, info.ID, colID, &colInfo.FieldType, distinct, false)
+					if err != nil {
+						return nil, errors.Trace(err)
+					}
+				}
+			}
+			if col != nil {
+				table.Columns = append(table.Columns, col)
+			} else {
+				log.Warnf("We can find column id %d in table %s now. It may be deleted.", colID, info.Name)
+			}
+		}
+	}
+	return table, nil
```

Comment:
I think that load table from storage in this way needs too much check. May be it is better to iterate through the table info to load stats.